### PR TITLE
first attemplt to implement hooks

### PR DESCRIPTION
**Pull request by JustFly1984**
# Please explain PR reason.
I've started to work on react hooks, anybody interested can join, or give advice from where I left it today.

### CODE REVIEW

**Comment by JustFly1984** [comment]
@uriklar can you please look?

File: packages/react-google-maps-api/src/GoogleMapNext.tsx [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+export function GoogleMap(props: GoogleMapProps): JSX.Element {
+  const [map, setMap] = React.useState<google.maps.Map | null>(null)
+
+  const [registeredEvents, setRedirectEvents] = React.useState<google.maps.MapsEventListener[]>([])
+
+  const mapRef = React.useRef<HTMLDivElement>(null)
+
+  React.useEffect(function onMount () {
```
Comment (uriklar) [inline packages/react-google-maps-api/src/GoogleMapNext.tsx]: Why do we need to separate mount and update to two separate useEffects functions?
This is a bit counter intuitive to the hooks way of things..

I think they could be united to one useEffect that maybe has a condition to handle mount specific logic

File: packages/react-google-maps-api/src/GoogleMapNext.tsx [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  onTiltChanged?: () => void
+  /** This event is fired when the map zoom property changes. */
+  onZoomChanged?: () => void
+  /** This callback is called when the component unmounts. It is called with the map instance. */
+  onLoad?: (map: google.maps.Map) => void | Promise<void>
+  /** This callback is called when the map instance has loaded. It is called with the map instance. */
+  onUnmount?: (map: google.maps.Map) => void | Promise<void>
+}
+
+export function GoogleMap(props: GoogleMapProps): JSX.Element {
```
Comment (uriklar) [inline packages/react-google-maps-api/src/GoogleMapNext.tsx]: I'm not sure if the GoogleMap component follows the exact same logic as the others, but when I initially tried implementing hooks the idea was creating a `useMapComponent` hook that accepts the component type, it's eventMap and updaterMap. 

Check out (not a working implementation, but the idea I think is pretty much there):
The "generic" hook:
https://github.com/JustFly1984/react-google-maps-api/blob/1b1940eef9a293c7e4743484bb75046e65bd8d5b/packages/react-google-maps-api/src/utils/use-map-component.ts

And a component that uses it:
https://github.com/JustFly1984/react-google-maps-api/blob/1b1940eef9a293c7e4743484bb75046e65bd8d5b/packages/react-google-maps-api/src/components/drawing/Rectangle.tsx

File: packages/react-google-maps-api/src/GoogleMapNext.tsx [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+      registeredEvents = applyUpdatersToPropsAndRegisterEvents({
+        updaterMap,
+        eventMap,
+        prevProps,
+        nextProps: props,
+        instance: map
+      })
+    }
+  }, [props])
```
[the remark this answers is not in the document]
Comment (vvmarulin) [reply packages/react-google-maps-api/src/GoogleMapNext.tsx]: And the same approach with `useRef` have to be used for caching of `registeredEvents`. React state has to be used to keep value that will be rendered. Saving `registeredEvents` in a state will make component renders one extra unnecessary time.

File: packages/react-google-maps-api/src/GoogleMapNext.tsx (hunk 2 above)
Comment (lluisrojass) [reply packages/react-google-maps-api/src/GoogleMapNext.tsx]: @uriklar curious, any reason this example omits the onLoad prop ? it's not due to any limitation is it?

**Pull request closed without merging**